docs/design/2020-05-08-cluster-index.md

```diff
@@ -0,0 +1,254 @@
+   IntValue() int64
+   // Next returns the minimum handle that is greater than this handle.
+   Next() Handle
+   // Equal returns if the handle equals to another handle, it panics if the types are different.
+   Equal(h Handle) bool
+   // Compare returns the comparison result of the two handles, it panics if the types are different.
+   Compare(h Handle) int
+   // Encoded returns the encoded bytes.
+   Encoded() []byte
+   // Len returns the length of the encoded bytes.
+   Len() int
+   // NumCols returns the number of columns of the handle,
+   NumCols() int
+   // EncodedCol returns the encoded column value at the given column index.
+   EncodedCol(idx int) []byte
+   // String implements the fmt.Stringer interface.
+   String() string
+}
+
+// IntHandle implements the Handle interface for the int64 type handle.
+type IntHandle int64
+
+// CommonHandle implements the Handle interface for non-int64 type handle.
+type CommonHandle struct {
+   encoded       []byte
+   colEndOffsets []uint16
+}
+
+// HandleMap is the map for Handle.
+type HandleMap struct {
+   ints map[int64]interface{}
+   strs map[string]strHandleVal
+}
+
+type strHandleVal struct {
+   h   Handle
+   val interface{}
+}
+```
+To extend the handle type, we need to abstract the handle to an interface that has the
+needed methods. 
+
+There are two concrete types: IntHandle and CommonHandle.
+
+If the primary key is a single column integer, the handle type would be IntHandle.
+Otherwise, the handle type would be CommonHandle.
+
+handles are used as map key in many places, one way to adapt to it is to use the encoded bytes as the key, but that would introduce memory allocation for IntHandle, so we use a HandleMap type that has two maps for the two handle types to avoid the memory allocation for IntHandle.
+
+Integer handle row key and index key formats are the same as the current implementation.
+We only need to care about the common handle type.
+
+### Codec
+
+#### common handle row key format
+ ```
+ t | {table_id} | _r | {common_handle}     // component
+ 1 | 8          | 2  | len(common_handle)  // byte size
+ ```
+The `common_handle` is the same as `index_column_values` in the index key.
+The row key length can be more than 19.
+
+#### non-unique index key format
+```
+ t | {table_id} | _i | {idx_col_vals}     | {common_handle}     // component
+ 1 | 8          | 2  | len(idx_col_vals) | len(common_handle) // byte size
+```
+Non-unique index value is unchanged.
+
+#### unique index value format
+```
+ {tailLen} | {common_handle_flag} | {common_handle_len} | {common_handle}    // component
+ 1         | 1                    | 2                   | len(common_handle) // byte size
+
+```
+The unique index key is unchanged.
+
+### Planner
+
+#### Handle Primary Key Index differently
+For common handle tables, treat the primary index access path is table path instead of index path, builds TableScan plan instead of IndexLookUp plan.
```

Comment:
also need to take care of handle table's "primary index" only cover a part of output columns (i.e. primary: a, b, c  but need a, b, c, d), so we need take cost to choose "join two tables(indexes)" or "do index lookup"?